[FEATURE] Traduire en anglais la page de connexion de Pix Orga (PIX-2221).

chrc opened the pull request:
## :unicorn: Problème
Dans Pix Orga, les textes de la page de connexion en français, et implémentés "en dur" dans le code.
On veut ajouter la version anglaise.

## :robot: Solution
* Externaliser dans des fichiers dédiés la version française et anglaise des textes de cette page
* Utiliser l'addon Ember d'internationalisation

## :rainbow: Remarques
> _Des infos supplémentaires, trucs et astuces ?_

## :100: Pour tester
Se connecter à Pix Orga 
* [en français](https://orga-pr2704.review.pix.fr/connexion?lang=fr)
* [en anglais](https://orga-pr2704.review.pix.fr/connexion?lang=en)

Scénarios:
* `is-only-accessible`: pas de connexion nécessaire, affiché au dessus du champ `email`
* `default`: faire renvoyer une erreur par `POST /api/token`
* `bad-request`: modifier la validation Joi de `POST /api/token` en ajoutant un champ obligatoire non envoyé par PixOrga
* aria-label "Voir le mot de passe" : inspecter la source (ex: anglais)
`<button class="pix-icon-button pix-icon-button--big pix-icon-button--dark-grey " aria-label="Show password" type="button">`
* utlisateur inexistant : `jaune.attend@example.org`
* mot de passe invalide : `lasttermsofservice@validated.net` / `foo`
* utilisateur non membre : `userpix1@example.net` / `pix123`
* invitation déjà acceptée
   * [français](https://orga-pr2704.review.pix.org/rejoindre?invitationId=10000000&code=SLEPTJ0312&lang=fr)     
   * [anglais](https://orga-pr2704.review.pix.org/rejoindre?invitationId=10000000&code=SLEPTJ0312&lang=en)    
* mot de passe à changer : requête ci-dessous + `sco.admin@example.net` / `pix123` 

Forcer l'utilisateur à changer son mot de passe
```
UPDATE
     "authentication-methods"
SET
    "authenticationComplement" = jsonb_set("authenticationComplement", '{shouldChangePassword}', '"true"')
WHERE 1=1
    AND "userId" IN (SELECT id FROM users u WHERE u.email = 'pixmaster@example.net')
;
```

chrc committed:
[ORGA] Add login page translations

chrc committed:
[ORGA] Update login templates with translation keys

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2704.review.pix.fr
- Orga: https://orga-pr2704.review.pix.fr
- Certif: https://certif-pr2704.review.pix.fr
- Admin: https://admin-pr2704.review.pix.fr
- API: https://api-pr2704.review.pix.fr/api/

 Please check it out!

chrc committed:
[API] Add code attribute to PasswordShouldChangeError

Add code 'SHOULD_CHANGE_PASSWORD' attribute to PasswordShouldChangeError to separate a simple http 401 error, and a 401 error of type PasswordShouldChangeError.
On the client side, this will display the appropriate error message.

chrc committed:
[ORGA] Manage lang during disconnection

chrc committed:
[ORGA] Hide recover link in en

chrc committed:
[ORGA] Use lang queryParams when not authenticated

octo-topi commented:
:chipotage:
>Les textes de la page de connection en français
>Les textes de la page de **connexion** en français

https://www.lalanguefrancaise.com/orthographe/connexion-ou-connection-orthographe/

AndreiaPena commented on orga/tests/integration/components/routes/login-form-test.js:
```diff
@@ -1,28 +1,38 @@
-  NOT_LINKED_ORGANIZATION_MSG: 'Vous ne pouvez pas vous connecter à PixOrga car vous n’êtes rattaché à aucune organisation. Contactez votre administrateur qui pourra vous inviter.',
-  INVALID_CREDENTIEL_MSG: 'L\'adresse e-mail et/ou le mot de passe saisis sont incorrects.',
-};
+import setupIntl from '../../../helpers/setup-intl';
```
Suggestion: Il existe un `setupIntlRenderingTest()` qui comprend le `setupRenderingTest()` et le `setupIntl()`.
Peut-être à privilégier ?

chrc replied:
Done

AndreiaPena commented on orga/translations/en.json:
```diff
@@ -309,6 +311,28 @@
+      "title": "Log in"
+    },
+    "login-form": {
+      "active-or-retrieve": "",
+      "email": "Email address",
+      "forgot-password": "Forgot your password?",
+      "forgotten-password-url": "https://app.pix.org/mot-de-passe-oublie?lang=en",
+      "invitation-already-accepted": "This invitation has already been accepted. Please log in or contact the administrator of your Pix Orga workspace.",
+      "is-only-accessible": "Pix Orga is only accessible to invited members. Please contact your Pix Orga administrator in order to be invited.",
+      "login": "Log in",
+      "only-for-admin": "",
```
**Remarque** : Tu avais fait le choix de ne pas mettre la traduction pour `"active-or-retrieve"` et  `"only-for-admin"` pour qu'ils n'apparaissent pas en anglais. Mais avec la nouvelle correction (displayRecoverLink) qui masque la div si on est en `en`, je propose qu'on remette les traductions ? Sauf s'il n'y a aucune chance que cette partie redevienne visible si on est en anglais ?

chrc replied:
Done

octo-topi commented:
**question**

Note: le comportement était déjà là avant la traduction

Si un utilisateur 
- n'est pas membre
- doit changer son mot de passe

On lui affiche le message suivant
![image](https://user-images.githubusercontent.com/56302715/111763794-d126ea00-88a2-11eb-8bee-0bb1c41be200.png)

Est-ce que cela ne serait pas plus logique de lui dire qu'il n'est pas membre de l'organisation ?

chrc commented:
> **question**
> 
> Note: le comportement était déjà là avant la traduction
> 
> Si un utilisateur
> 
> * n'est pas membre
> * doit changer son mot de passe
> 
> On lui affiche le message suivant
> ![image](https://user-images.githubusercontent.com/56302715/111763794-d126ea00-88a2-11eb-8bee-0bb1c41be200.png)
> 
> Est-ce que cela ne serait pas plus logique de lui dire qu'il n'est pas membre de l'organisation ?

@octo-topi  Normalement il y a un message d'erreur spécifique pour les non-membres d'une organisation. 
Quel est le user que tu as utilisé pour faire ce test (lasttermsofservice@validated.net/pix123 ne correspond pas exactement à ce profil) ?

**Update :** Effectivement, il y a un problème de "persistance" de l'ancien message d'erreur sur la page, malgré le HTTP code 403 renvoyé par l'API ! 

**DONE**

octo-topi commented:
**AFH**
Je n'arrive pas à tester la traduction de `is-only-accessible,` quel est le scénario ?
`"is-only-accessible": "Pix Orga is only accessible to invited members. Please contact your Pix Orga administrator in order to be invited.",`

chrc committed:
[ORGA] Add @tracked to login-form errorMessage property

chrc commented:
> is-only-accessible

@octo-topi Cela est utilisé sur la double mire de connexion (cf. https://github.com/1024pix/pix/blob/fd90764a343bb76b83c408c61929be6cb7886386/orga/app/components/routes/login-or-register.hbs#L27).

AndreiaPena commented:
@octo-topi @chrc  `is-only-accessible` c'est le message dans la page de connexion non ? 🧐
<img width="521" alt="Capture d’écran 2021-03-19 à 15 02 16" src="https://user-images.githubusercontent.com/58915422/111792040-1d365680-88c4-11eb-9406-d588d2dc90dc.png">

octo-topi commented on api/tests/acceptance/application/authentication-controller_test.js:
```diff
@@ -84,13 +84,12 @@ describe('Acceptance | Controller | authentication-controller', () => {
+          code: 'SHOULD_CHANGE_PASSWORD',
+          detail: 'Erreur, vous devez changer votre mot de passe.',
+          status: '401',
+          title: 'PasswordShouldChange',
```
**chipotage**
Préciser qui fait quoi
```suggestion
          title: 'userShouldChangePassword',
```

chrc replied:
Je ne vais pas faire de changement dans ce ticket.
Mais si je veux rester au plus proche de l'[ADR (point 5)](https://jsonapi.org/format/#errors), je rechercherais :
* _a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization_
* Un focus sur  le couple HTTP statusCode et `error.code` pour gérer la traduction des  messages d'erreur (côté front)

octo-topi commented on orga/app/components/routes/login-form.hbs:
```diff
@@ -58,22 +55,25 @@
+        </a>
       </div>
-      {{#unless @isWithInvitation}}
+      {{#if this.displayRecoverLink}}
```
**chipotage**
```suggestion
      {{#if this.displayRecoveryLink}}
```
ou
```suggestion
      {{#if this.displayLinkToRecoverAccess}}
```

chrc replied:
Done

octo-topi commented:
Pas d'ADR pour finir :cry: ?
Y'a du sujet avec gestion erreur API et i18n back !
Si tu veux de l'aide, je peux prendre

chrc commented:
> Pas d'ADR pour finir 😢 ?
> Y'a du sujet avec gestion erreur API et i18n back !
> Si tu veux de l'aide, je peux prendre

@octo-topi : En fait, tout était déjà présent côté API, et donc dans l'ADR ;-)

octo-topi commented:
> > is-only-accessible
> 
> @octo-topi Cela est utilisé sur la double mire de connexion (cf.
> 
> https://github.com/1024pix/pix/blob/fd90764a343bb76b83c408c61929be6cb7886386/orga/app/components/routes/login-or-register.hbs#L27
> ).

Comme vu avec @AndreiaPena, la traduction est 
- affichée lors de la connexion (traduction testée OK)
- cachée dans la double mire (non-régression testée OK)

octo-topi commented on orga/app/components/routes/login-form.js:
```diff
@@ -49,16 +60,16 @@ export default class LoginForm extends Component {
   _authenticate(password, email) {
     const scope = 'pix-orga';
+
+    this.isErrorMessagePresent = false;
```
**chipotage**
[La doc ](https://api.emberjs.com/ember/3.5/classes/@ember%2Fobject%2Fcomputed/methods/alias?anchor=empty) indique qu'il est possible de déclarer une variable comme lié à une autre, et qui vaut true/false si l'autre est renseignée.

Mais je n'ai pas réussi à le faire fonctionner :cry:

octo-topi commented on orga/app/routes/application.js:
```diff
@@ -47,8 +47,8 @@ export default class ApplicationRoute extends Route.extend(ApplicationRouteMixin
-  _setLocale() {
-    const locale = get(this.currentUser, 'prescriber.lang', DEFAULT_LOCALE);
+  _setLocale(lang = null) {
+    const locale = lang || get(this.currentUser, 'prescriber.lang', DEFAULT_LOCALE);
```
**chipotage**
S'inspirer de la solution utilisée [pix-app](http://github.com/1024pix/pix/blob/dev/mon-pix/app/routes/application.js#L74-L74) pour rendre le code plus expressif

octo-topi commented on orga/app/routes/application.js:
```diff
@@ -35,7 +35,7 @@ export default class ApplicationRoute extends Route.extend(ApplicationRouteMixin
     await this._loadCurrentUser();
     await this._updatePrescriberLocale(lang);
-    this._setLocale();
+    this._setLocale(lang);
```
**question**
Une fonction `_get` qui fait un `_set :thinking:

octo-topi commented on orga/app/routes/application.js:
```diff
@@ -35,7 +35,7 @@ export default class ApplicationRoute extends Route.extend(ApplicationRouteMixin
   async _getUserAndLocale(lang = null) {
```
**question**
Une fonction `_get` qui fait un `_set` et un `_update` :thinking:

octo-topi replied:
**chipotage**
```suggestion
  async _getUserAndLocale(langFromQueryParam = null) {
```

octo-topi commented on orga/app/services/url.js:
```diff
@@ -18,7 +19,8 @@ export default class Url extends Service {
   get homeUrl() {
-    return this.definedHomeUrl;
+    const currentLanguage = this.intl.t('current-lang');
+    return `${this.definedHomeUrl}?lang=${currentLanguage}`;
```
**chipotage**
Le `defined` n'apporte pas de valeur. 
Je propose  de l'enlever ou de remplacer par
```suggestion
    homeUrlWithoutLang = ENV.rootURL;
    (..)
    get homeUrl() {
        homeUrlWithoutLang = ENV.rootURL;
        (..)
       return `${this.homeUrlWithoutLang}?lang=${currentLanguage}`;
```
